Custom Jobs (#494)
timescale/docs.timescale.com-content: api.md

performance on older chunks. Each policy is assigned a scheduled job
which will be run in the background to enforce it.

## add_job() :community_function: [](add_job)

Register a custom procedure or function to be scheduled by our automation framework.

#### Required Arguments [](add_job-required-arguments)

|Name|Description|
|---|---|
| `proc` | (REGPROC) Name of the function or procedure to register as job|
| `schedule_interval` | (INTERVAL) Interval between executions of this job|

#### Optional Arguments [](add_job-optional-arguments)

|Name|Description|
|---|---|
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|

> | `initial_start` | (TIMESTAMPTZ) Time of first execution of job. |

Reviewer: Minor nit:  Our standard is not to include "periods" at the end of the descriptions (unless there is a multiple sentence description).  This uses a mix, cut drop the period?

Author: OK that's not applied consistently then

| `scheduled` | (BOOLEAN) Set to `FALSE` to exclude this job from scheduling. Defaults to `TRUE`. |

#### Returns [](add_job-returns)

|Column|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id |

#### Sample Usage [](add_job-examples)

```sql

> SELECT add_job('custom_proc','1h', config:='{"type":"procedure"}'::jsonb);

Reviewer 2: It would be more helpful if we first show the create procedure/function stmt for a function that takes an argument. And then show how to add  a job for that procedure/function as the next step. 

Also need to mention what is a valid config and the minimum information for the jsonb config. Can it be NULL for instance? Or do we always have to pass in type information?

Reviewer: Agree.  I think here we will largely point to the "using custom jobs" section of the docs which will go into depth (forthcoming), but it would be nice to have a very single example here where you just show very simple definition of the UDF, then use that example to `add_job`.

```

Register the procedure custom_proc to be run every hour.

## delete_job() :community_function: [](delete_job)

Delete a job registered with the automation framework.
This works for custom jobs as well as policies.

If the job is currently running the process will be terminated.

#### Required Arguments [](delete_job-required-arguments)

|Name|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id |

#### Sample Usage [](delete_job-examples)

```sql
SELECT delete_job(1000);
```

Delete the job with the job id 1000.

## run_job() :community_function: [](run_job)

Run a previously registered job in the current session.
This works for custom jobs as well as policies.

#### Required Arguments [](run_job-required-arguments)

|Name|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id |

#### Sample Usage [](run_job-examples)

```sql

> CALL run_job(1000);

Reviewer 2: would be good to give a bit of context and elaborate a bit.
1. select * from timescaledb_information.xxxx and choose a specific policy , e.g.  drop_chunks or refresh policy.
2. then CALL run_job( <job_id>);

Reviewer: Clarify use of `CALL` (and perhaps one line why -- this is something new that didn't appear in 1.x).

performance on older chunks. Each policy is assigned a scheduled job
which will be run in the background to enforce it.

## add_job() :community_function: [](add_job)

Register a custom procedure or function to be scheduled by our automation framework.
Please read the [instructions][using-custom-jobs] for more details including
multiple example jobs.

#### Required Arguments [](add_job-required-arguments)

|Name|Description|
|---|---|
| `proc` | (REGPROC) Name of the function or procedure to register as job|

> | `schedule_interval` | (INTERVAL) Interval between executions of this job|

Reviewer 2: If it is not a scheduled job, wouldn't this be NULL? Shouldn't this be optional in that case?

Author: the column is not null so it is required, it's in the design doc

#### Optional Arguments [](add_job-optional-arguments)

|Name|Description|
|---|---|
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|
| `initial_start` | (TIMESTAMPTZ) Time of first execution of job |
| `scheduled` | (BOOLEAN) Set to `FALSE` to exclude this job from scheduling. Defaults to `TRUE`. |

#### Returns [](add_job-returns)

|Column|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id |

#### Sample Usage [](add_job-examples)

```sql
CREATE OR REPLACE PROCEDURE custom_job(job_id int, config jsonb) LANGUAGE PLPGSQL AS
$$
BEGIN
  RAISE NOTICE 'Executing job % with config %', job_id, config;
END
$$;

SELECT add_job('custom_proc','1h');
```

> Register the procedure custom_proc to be run every hour.

Reviewer: Pointer to file in "Using" section for more examples.

```

---
## add_job() :community_function: [](add_job)

Register an action to be scheduled by our automation framework.
Please read the [instructions][using-actions] for more details including
multiple example actions.

#### Required Arguments [](add_job-required-arguments)

|Name|Description|
|---|---|
| `proc` | (REGPROC) Name of the function or procedure to register as job|
| `schedule_interval` | (INTERVAL) Interval between executions of this job|

#### Optional Arguments [](add_job-optional-arguments)

|Name|Description|
|---|---|
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|
| `initial_start` | (TIMESTAMPTZ) Time of first execution of job |
| `scheduled` | (BOOLEAN) Set to `FALSE` to exclude this job from scheduling. Defaults to `TRUE`. |

#### Returns [](add_job-returns)

|Column|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id |

#### Sample Usage [](add_job-examples)

```sql

> CREATE OR REPLACE PROCEDURE custom_job(job_id int, config jsonb) LANGUAGE PLPGSQL AS

Reviewer: ```suggestion
CREATE OR REPLACE PROCEDURE user_defined_action(job_id int, config jsonb) LANGUAGE PLPGSQL AS
```

$$
BEGIN

>   RAISE NOTICE 'Executing job % with config %', job_id, config;

Reviewer: ```suggestion
  RAISE NOTICE 'Executing action % with config %', job_id, config;
```

END
$$;

> SELECT add_job('custom_proc','1h');

Reviewer: ```suggestion
SELECT add_job('user_defined_action','1h');
```

```

> Register the procedure `custom_proc` to be run every hour.

Reviewer: ```suggestion
Register the procedure `user_defined_action` to be run every hour.
```

## delete_job() :community_function: [](delete_job)

Delete a job registered with the automation framework.
This works for user-defined actions as well as policies.

> If the job is currently running the process will be terminated.

Reviewer: ```suggestion
If the job is currently running, the process will be terminated.
```

---

## alter_job() :community_function: [](alter_job)

Policy jobs are scheduled to run periodically via a job run in a background

> worker. You can change the schedule using `alter_job`. To alter an

Reviewer 2: This specifically refers to schedule. But we should be able to change any attribute. Need changes to reflect that.

existing job, you must refer to it by `job_id`. The `job_id` which implements a
given policy and its current schedule can be found in views in the

> `timescaledb_information` schema corresponding to different types of policies or

Reviewer 2: we don't seem to have policy specific views anymore. Are you planning to add one which shows all the polices? This part needs an update.

| `if_exists` | (BOOLEAN)  Set to true to avoid throwing an error if the job does not exist, a notice will be issued instead. Defaults to false. |

#### Returns [](alter_job-returns)

|Column|Description|
|---|---|
| `job_id` | (INTEGER) the id of the policy job being modified |
| `schedule_interval` | (INTERVAL)  The interval at which the job runs |
| `max_runtime` | (INTERVAL) The maximum amount of time the job will be allowed to run by the background worker scheduler before it is stopped |
| `max_retries` | (INTEGER)  The number of times the job will be retried should it fail |
| `retry_period` | (INTERVAL) The amount of time the scheduler will wait between retries of the job on failure |
| `scheduled` | (BOOLEAN)  True if this job will be executed by the TimescaleDB scheduler. |
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|
| `next_start` | (TIMESTAMPTZ) The next time at which to run the job. |

#### Sample Usage [](alter_job-examples)

```sql
SELECT alter_job(1000, schedule_interval => INTERVAL '2 days');
```

> reschedules the job with id 1000 so that it runs every two days.

Reviewer 2: reschedules ->Reschedules

```sql
SELECT alter_job(job_id, schedule_interval => INTERVAL '5 minutes')
FROM timescaledb_information.continuous_aggregate_stats
WHERE view_name = 'conditions_agg'::regclass;
```

> reschedules the continuous aggregate job for the `conditions_agg` view so that it runs every five minutes.

Reviewer 2: Same here. capitalize reschedules.

interval or reorder by a different index, you'll need to remove the policy and
add a new one.

#### Required Arguments [](alter_job-required-arguments)

|Name|Description|
|---|---|
| `job_id` | (INTEGER) the id of the policy job being modified |

#### Optional Arguments [](alter_job-optional-arguments)

|Name|Description|
|---|---|
| `schedule_interval` | (INTERVAL)  The interval at which the job runs |
| `max_runtime` | (INTERVAL) The maximum amount of time the job will be allowed to run by the background worker scheduler before it is stopped |
| `max_retries` | (INTEGER)  The number of times the job will be retried should it fail |
| `retry_period` | (INTERVAL) The amount of time the scheduler will wait between retries of the job on failure |

> | `scheduled` | (BOOLEAN)  Set to `FALSE` to exclude this job from being run as background job. |

Reviewer: For all boolean optional arguments, specify the default?  (It does on some but not all)

Author: This is optional arguments section where default is NULL

| `next_start` | (TIMESTAMPTZ) The next time at which to run the job. The job can be paused by setting this value to 'infinity' (and restarted with a value of now()). |
| `if_exists` | (BOOLEAN)  Set to true to avoid throwing an error if the job does not exist, a notice will be issued instead. Defaults to false. |

#### Returns [](alter_job-returns)

|Column|Description|
|---|---|
| `job_id` | (INTEGER) the id of the policy job being modified |
| `schedule_interval` | (INTERVAL)  The interval at which the job runs |
| `max_runtime` | (INTERVAL) The maximum amount of time the job will be allowed to run by the background worker scheduler before it is stopped |
| `max_retries` | (INTEGER)  The number of times the job will be retried should it fail |
| `retry_period` | (INTERVAL) The amount of time the scheduler will wait between retries of the job on failure |
| `scheduled` | (BOOLEAN)  True if this job will be executed by the TimescaleDB scheduler. |
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|
| `next_start` | (TIMESTAMPTZ) The next time at which to run the job. |

#### Sample Usage [](alter_job-examples)

```sql

> SELECT alter_job(job_id, schedule_interval => INTERVAL '2 days')

Reviewer 2: The views used by these examples have been dropped. So needs changes.

FROM timescaledb_information.reorder_policies

> WHERE hypertable = 'conditions'::regclass;

Reviewer: Would this query work without the explicit `::regclass` cast/specification in the WHERE clause?  I realize this is correct / "more" correct, but the typing gets slightly more confusing for your average user.  So perhaps would be preferable to show examples without it, _assuming_ those work.

Author: no

---

## alter_job() :community_function: [](alter_job)

> Policy jobs are scheduled to run periodically via a job run in a background
> worker. You can change the schedule using `alter_job_schedule`. To alter an
> existing job, you must refer to it by `job_id`. The `job_id` which implements a
> given policy and its current schedule can be found in views in the
> `timescaledb_information` schema corresponding to different types of policies or

Reviewer: How about this for the next paragraph:
```suggestion
Actions scheduled via TimescaleDB's automation framework run periodically in a
background worker. You can change the schedule of their execution using `alter_job`.
To alter an existing job, you must refer to it by `job_id`.

The `job_id` which executes a given action and its current schedule can be found
either in the `timescaledb_information.job_stats` view, which lists information 
about every scheduled action, as well as in corresponding views in the
`timescaledb_information` schema provided for built-in policies (e.g., related
to data retention, compression, and continuous aggregates).  The `job_stats` view
additionally contains information about when each job was last run and other
useful statistics for deciding what the new schedule should be.
```

|Name|Description|
|---|---|
| `job_id` | (INTEGER) the id of the policy job being modified |

#### Optional Arguments [](alter_job-optional-arguments)

|Name|Description|
|---|---|
| `schedule_interval` | (INTERVAL)  The interval at which the job runs |
| `max_runtime` | (INTERVAL) The maximum amount of time the job will be allowed to run by the background worker scheduler before it is stopped |
| `max_retries` | (INTEGER)  The number of times the job will be retried should it fail |
| `retry_period` | (INTERVAL) The amount of time the scheduler will wait between retries of the job on failure |
| `scheduled` | (BOOLEAN)  Set to `FALSE` to exclude this job from being run as background job. |
| `config` | (JSONB) Job-specific configuration (this will be passed to the function when executed)|
| `next_start` | (TIMESTAMPTZ) The next time at which to run the job. The job can be paused by setting this value to 'infinity' (and restarted with a value of now()). |
| `if_exists` | (BOOLEAN)  Set to true to avoid throwing an error if the job does not exist, a notice will be issued instead. Defaults to false. |

#### Returns [](alter_job-returns)

|Column|Description|
|---|---|

> | `job_id` | (INTEGER) the id of the policy job being modified |

Reviewer: ```suggestion
| `job_id` | (INTEGER) the id of the job being modified |
```
